Added HTTPS,HTTProtocol, part of #24 (#74)
The-Mycelium-Network/webworkbook: content/topics/protocols/index.md

---
title: What are protocols?
keywords: internet protocols, types of internet protocols, internet protocols explained
description: A collection of informational pieces that explain internet protocols.
menu:
  main:
    parent: topics
    name: Protocols
---

> ## What are protocols exactly?

Reviewer: Agreed, the "exactly" is an interesting resource in text, but it kinda feels that it would become after an introductory bit, often polemicizing... like: "Some people say this, some people say that. So, what are protocols exactly?"

> ## What are protocols exactly?

Reviewer: ```suggestion
## What are protocols?

```

> ## What are protocols exactly?

Reviewer: ```suggestion
## What are protocols?

```

SMTP servers handle the sending, receiving, and relaying of emails that will later be downloaded via a receiving protocol. A good metaphor for SMTP servers is real-life post offices. When you send a letter, your local post office processes your correspondence and decides which post office to forward the letter to. After reaching the recipient's post office, the letter will be ready to be retrieved by your local mailman and delivered to your address.

**References:**

- [MDN Web Docs Glossary](https://developer.mozilla.org/en-US/docs/Glossary/SMTP)
- [Wikipedia](https://en.wikipedia.org/wiki/Simple_Mail_Transfer_Protocol)
- [GeeksforGeeks](https://www.geeksforgeeks.org/simple-mail-transfer-protocol-smtp/)

**See also:**

- [Postmark: Everything you need to know about SMTP](https://postmarkapp.com/guides/everything-you-need-to-know-about-smtp)
- Videos:
  - [Eye on Tech: What is SMTP (Simple Mail Transfer Protocol)?](https://www.youtube.com/watch?v=AoqvWgazf50)
  - [TECHNOPHILE: What Is SMTP ? How it Works !](https://www.youtube.com/watch?v=O-Xho7gEWj0)

> ### POP3

Reviewer: ```suggestion
### POP3 (Post Office Protocol 3)

```

Consistency.

**References:**

- [Techtarget](https://www.techtarget.com/whatis/definition/POP3-Post-Office-Protocol-3)
- [Hostinger](https://www.hostinger.com/tutorials/email/pop3-imap-smtp-protocols-explained-ports)

### IMAP (Internet Message Access Protocol)

This is one of a few methods used to retrieve e-mails. This method makes it possible for the user to access messages from any device and e-mails will also be synchronized across all devices that are being used. IMAP stores e-mails on an external server allowing remote access. The same goes for outgoing messages. Messages are only downloaded once you click on them and attachments are not automatically downloaded either, making IMAP fast, efficient and versatile.

**References:**

- [WhatIsMyIPAddress: What is a IMAP?](https://whatismyipaddress.com/imap)
- [JavaTpoint: IMAP Protocol](https://www.javatpoint.com/imap-protocol)
- [GeeksforGeeks: Internet Message Access Protocol (IMAP)](https://www.geeksforgeeks.org/internet-message-access-protocol-imap/)

> ### UDP

Reviewer: ```suggestion
### UDP (User Datagram Protocol)

```

Title consistency.

UDP stands for User Datagram Protocol. UDP is specifically chosen for time-sensitive applications like online gaming, video playback, and Domain Name System (DNS) lookups. UDP uses a [connectionless communication](https://en.wikipedia.org/wiki/Connectionless_communication) model, which speeds up communication between client and server. Unlike Transmission Control Protocol(TCP), the order of packets (called datagrams) is not guaranteed, nor are there any checks to ensure packets are delivered. Because a formal connection is not made, UDP can leave one open to _[DDoS attacks](https://www.fortinet.com/resources/cyberglossary/ddos-attack)_ through UDP [flood attacks](https://www.cloudflare.com/learning/ddos/udp-flood-ddos-attack/).

**References:**

> - [MDN-UDP](https://developer.mozilla.org/en-US/docs/Glossary/UDP)
> - [Cloudflare-UDP](https://www.cloudflare.com/learning/ddos/glossary/user-datagram-protocol-udp/)

Reviewer 2: For consistency:

```suggestion
- [MDN Web Docs: UDP](https://developer.mozilla.org/en-US/docs/Glossary/UDP)
- [Cloudflare: UDP](https://www.cloudflare.com/learning/ddos/glossary/user-datagram-protocol-udp/)
```

### FTP (File Transfer Protocol)

File Transfer Protocol is used to allow computers to transfer files from one system to another using the internet. FTP works by opening a connection between 2 parties, which is made via two distinct communication channels: the control channel handles FTP commands and responses; while the data channel is where the sharing happens once the server has granted you access.

The FTP is insecure by current standards and it's considered a legacy protocol, the main issue being the fact that it does not encrypt the data sent, allowing data interception. There are protocols more appropriate to safely transfer data, such as:

> - FTPS (FTP Secure), which adds a security layer on top of FTP using cryptographic protocols like SSL or TLS;
> - SFTP (SSH File Transfer Protocol), which runs over the SSH Protocol and provides encrypted data communication.

Reviewer 2: For consistency:

```suggestion
- FTP Secure (FTPS), which adds a security layer on top of FTP using cryptographic protocols like SSL or TLS;
- SSH File Transfer Protocol (SFTP) runs over the SSH Protocol and provides encrypted data communication.

```

**References:**

- [MDN Web Docs Glossary: FTP](https://developer.mozilla.org/en-US/docs/Glossary/FTP)

**See also:**

- [Hostinger Tutorials: What is FTP - FTP Explained for Beginners](https://www.hostinger.com/tutorials/what-is-ftp)
- Videos:
  - [CertBros: FTP Explained | File Transfer Protocol | Cisco CCNA 200-301](https://www.youtube.com/watch?v=L9aZpg0ip70)

### FTPS (File Transfer Protocol Secure)

> FTP Secure is used to securely transfer data between users, companies or clients. It's an extension of the File Transfer Protocol (FTP) which adds support for a security layer. FTPS uses cryptographic protocols like SSL (Secure Sockets Layer) or TLS (Transport Layer Security), although the former is now deprecated due to security vulnerabilities.

Reviewer 2: > FTP Secure is used to securely transfer data between

Maybe we should rather say:

> FTP Secure is used to securely transfer data between computer systems.

> In a nutshell, the data gets encrypted before it's sent, preventing it from being intercepted and read by third parties. FTP make use of client certification and server identities when transferring the data from point A to point B, and various security measures and encryption strategies are used to connect with foreign servers and verify authenticity before sending any data.

Reviewer 2: ```suggestion
In a nutshell, the data gets encrypted before it's sent, preventing it from being intercepted and read by third parties. FTPS uses client certification and server identities when transferring the data from point A to point B, and various security measures and encryption strategies are used to connect with foreign servers and verify authenticity before sending any data.

```

**References:**

- [Precisely glossary: FTPS](https://www.precisely.com/glossary/ftps)
- [Wikipedia: FTPS](https://en.wikipedia.org/wiki/FTPS)

### SFTP (SSH File Transfer Protocol)

SFTP functions pretty much the same as FTPS protocol, both being protocols used to transfer files safely through an encrypted connection. The difference is the encryption type and the authentication process used. While FTPS just adds a security layer around the legacy FTP protocol, SFTP is an extension of the SSH protocol, which allows remote machines to verify each other before connecting and transferring data.

SFTP is quickly taking the place of FTP and FTPS. The reason for this is that SFTP is more secure, more reliable and easier to configure. This means that there are not many reasons to use any other protocol in its place. SFTP is protected against [man-in-the-middle-attacks](https://www.ssh.com/academy/attack/man-in-the-middle) and even [password sniffing](https://www.ssh.com/academy/attack/password-sniffing).

**References:**

> - [SFTP-SSH Academy](https://www.ssh.com/academy/ssh/protocol)

Reviewer: ```suggestion
- [SSH Academy: SFTP File Transfer Protocol - get SFTP client & server](https://www.ssh.com/academy/ssh/sftp)

```

Wrong link.

- [GeeksforGeeks: Difference between FTPS and SFTP](https://www.geeksforgeeks.org/difference-between-ftps-and-sftp/)
- [SmartFile: The Difference Between SFTP & FTPS and Why It Matters](https://www.smartfile.com/blog/difference-between-sftp-ftps/)

### SSH (Secure Shell)

SSH provides a secure method to remotely login from one computer to another, with several alternative options for authentication, along with a strong encryption to protect the integrity of the communications. SSH is used in conjunction with the SFTP protocol, due to the high level of security it provides. This protocol is used in a corporate network in the following ways:

> In corporate networks, SSH protocol can be used to guarantee secure access for users and allow issuing remote commands. A strong authentication protocol is also key for running automated processes and file transfers without risking an invasion or data breach.

Reviewer: ```suggestion
In corporate networks, SSH protocol can be used to guarantee secure access for users and allow issuing of remote commands. A strong authentication protocol is also key for running automated processes and file transfers without risking an invasion or data breach.

```

**References:**

> - [SSH-SSH Academy](https://www.ssh.com/academy/ssh/protocol)

Reviewer: ```suggestion
- [SSH Academy: SSH Protocol – Secure Remote Login and File Transfer](https://www.ssh.com/academy/ssh/protocol)

```

> ### SSL **Secure Sockets Layer**

Reviewer: ```suggestion
### SSL (Secure Sockets Layer)

```

{{< youtube id="ZghMPWGXexs&ab" title="A video explaining SSL encryption and decryption" >}}

Secure socket layer (SSL) is a networking protocol that keeps your internet connection secure and ensures that the data being transferred remains private. This protocol's main use is securing the connection/communication between your browser and a web server. It does this by scrambling the data and making it impossible to read to anyone trying to intercept it. SSL uses cryptographic keys to encrypt information; a public key encrypts the data, and a private key decrypts it (making it readable). An SSL certificate enables websites to move from HTTP to HTTPS.

**References:**

> - [TechTarget](htttps://www.techtarget.com/searchsecurity/definition/Secure-Sockets-Layer-SSL)

Reviewer: ```suggestion
- [TechTarget: SSL (secure sockets layer)](https://www.techtarget.com/searchsecurity/definition/Secure-Sockets-Layer-SSL)

```

Too many t's ;)

> - [CloudFlare](https://www.cloudflare.com/learning/ssl/what-is-ssl/)

Reviewer: ```suggestion
- [CloudFlare: What is SSL? | SSL definition](https://www.cloudflare.com/learning/ssl/what-is-ssl/)

```

> - [CSoOnline](https://www.csoonline.com/article/3246212/what-is-ssl-how-ssl-certificates-enable-encrypted-communication.html)

Reviewer: ```suggestion
- [CSoOnline: What is SSL? How SSL certificates enable encrypted communication](https://www.csoonline.com/article/3246212/what-is-ssl-how-ssl-certificates-enable-encrypted-communication.html)

```

> ### HTTP **Hyper Text Transfer Protocol**

Reviewer: ```suggestion
### HTTP (Hyper Text Transfer Protocol)

```

Consistency.

> HTTP is the protocol of the internet,it is responsible for communication between web severs and clients. It is simple and designed to be human readable. This protocol is also extensible,thanks to HTTP [headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers), it is easy to expand and experiment with.

Reviewer: ```suggestion
HTTP is the protocol of the internet and it is responsible for communication between web servers and clients. It is simple and designed to be human-readable. This protocol is also extensible, thanks to HTTP [headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers), it is easy to expand and experiment with.

```

> HTTP is stateless,this means that each request that the server receives is independent and not linked to any prior requests.

Reviewer: ```suggestion
HTTP is stateless, this means that each request that the server receives is independent and not linked to any prior requests.

```

> HTTP and all extended protocols based on HTTP are established on the [request and response model](https://medium.com/@reemshakes/http-request-response-flow-for-dummies-a0f52af83af3),meaning the client sends a _request_ to the server,and the server sends back a _response_,if the request fails the response will be an error code. A well know error code is 404 - Not Found. For the full list of HTTP error codes click [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status)

Reviewer 2: ```suggestion
HTTP and all extended protocols based on HTTP are established on the [request and response model](https://medium.com/@reemshakes/http-request-response-flow-for-dummies-a0f52af83af3), meaning the client sends a `request` to the server, and the server sends back a `response` if the request fails the response will be an error code. A well-known error code is 404 - Not Found. For the full list of HTTP error codes, click [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status)

```

**References:**

> [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Overview#http_is_stateless_but_not_sessionless)
> [Traversy Media - HTTP Crash Course](https://www.youtube.com/watch?v=iYM2zFP3Zn0&ab_channel=TraversyMedia)
> [OmniSecu Tutorials](https://www.omnisecu.com/tcpip/http-hypertext-transfer-protocol-request-response.php)

Reviewer: ```suggestion
- [MDN Web Docs: An overview of HTTP](https://developer.mozilla.org/en-US/docs/Web/HTTP/Overview)
- [Traversy Media: HTTP Crash Course & Exploration](https://www.youtube.com/watch?v=iYM2zFP3Zn0&ab_channel=TraversyMedia)
- [OmniSecu Tutorials: Hypertext Transfer Protocol (HTTP) - Request and Response](https://www.omnisecu.com/tcpip/http-hypertext-transfer-protocol-request-response.php)

```

> ### HTTPS

Reviewer: ```suggestion
### HTTPS (Hyper Text Transfer Protocol Secure)

```

{{< youtube id="nOmT_5hqgPk" title="HTTP vs HTTPS">}}

> HTTPS is a secure version of the HTTP protocol that is encrypted and authenticated by the SSL/TLS protocol. This connection is secure thus allowing clients to safely exchange sensitive data with a server,like banking and credit card details

Reviewer: ```suggestion
HTTPS is a secure version of the HTTP protocol that is encrypted and authenticated by the SSL/TLS protocol. This connection is secure, thus allowing clients to safely exchange sensitive data with a server, like banking and credit card details.

```

**References:**

> [SSL.com](https://www.ssl.com/faqs/what-is-https/)
> [Tech Target](https://www.techtarget.com/searchsoftwarequality/definition/HTTPS)
> [Cloudflare.com](https://www.cloudflare.com/en-gb/learning/ssl/what-is-https/)

Reviewer: ```suggestion
- [SSL.com: What is HTTPS?](https://www.ssl.com/faqs/what-is-https/)
- [Tech Target: Hypertext Transfer Protocol Secure (HTTPS)](https://www.techtarget.com/searchsoftwarequality/definition/HTTPS)
- [Cloudflare: What is HTTPS?](https://www.cloudflare.com/en-gb/learning/ssl/what-is-https/)

```

> ### DNS **Domain Name System**

Reviewer: ```suggestion
### DNS (Domain Name System)

```

- The primary server is used for all administrative tasks.
- The secondary is the backup server.
- The caching server only performs cached-query services.

**References:**

- [DNSimple: How DNS Works](https://howdns.works/)
- [NS1: What is DNS? DNS Explained](https://ns1.com/resources/what-is-dns)
- [Microsoft Learn: DNS Servers](https://learn.microsoft.com/en-us/windows/win32/dns/dns-servers)

### DNS **Domain Name System**

{{< youtube id="5o8CwafCxnU&t=2s" title="A video explaining IP addresses and DNS" >}}

A standard protocol that works like a phonebook - it manages the mapping between names (websites) and numbers (IP addresses). This means that you don't need to remember complicated IP addresses, because DNS translates your domain name (URL) to the corresponding address. There are 3 types of DNS servers, namely: primary, secondary and caching servers.

- The primary server is used for all administrative tasks.
- The secondary is the backup server.
- The caching server only performs cached-query services.

**References:**

- [DNSimple: How DNS Works](https://howdns.works/)
- [NS1: What is DNS? DNS Explained](https://ns1.com/resources/what-is-dns)

> - [Microsoft Learn: DNS Servers](https://learn.microsoft.com/en-us/windows/win32/dns/dns-servers)

Reviewer 2: Whoops, this is duplicated.

> 
> ### DNS **Domain Name System**
> 
> {{< youtube id="5o8CwafCxnU&t=2s" title="A video explaining IP addresses and DNS" >}}
> 
> A standard protocol that works like a phonebook - it manages the mapping between names (websites) and numbers (IP addresses). This means that you don't need to remember complicated IP addresses, because DNS translates your domain name (URL) to the corresponding address. There are 3 types of DNS servers, namely: primary, secondary and caching servers.
> 
> - The primary server is used for all administrative tasks.
> - The secondary is the backup server.
> - The caching server only performs cached-query services.
> 
> **References:**
> 
> - [DNSimple: How DNS Works](https://howdns.works/)
> - [NS1: What is DNS? DNS Explained](https://ns1.com/resources/what-is-dns)
> - [Microsoft Learn: DNS Servers](https://learn.microsoft.com/en-us/windows/win32/dns/dns-servers)

Reviewer: ```suggestion

```

The FTP is insecure by current standards and it's considered a legacy protocol, the main issue being the fact that it does not encrypt the data sent, allowing data interception. There are protocols more appropriate to safely transfer data, such as:

- FTPS (FTP Secure), which adds a security layer on top of FTP using cryptographic protocols like SSL or TLS;
- SFTP (SSH File Transfer Protocol), which runs over the SSH Protocol and provides encrypted data communication.

**References:**

- [MDN Web Docs Glossary: FTP](https://developer.mozilla.org/en-US/docs/Glossary/FTP)

**See also:**

- [Hostinger Tutorials: What is FTP - FTP Explained for Beginners](https://www.hostinger.com/tutorials/what-is-ftp)
- Videos:
  - [CertBros: FTP Explained | File Transfer Protocol | Cisco CCNA 200-301](https://www.youtube.com/watch?v=L9aZpg0ip70)

### FTPS (File Transfer Protocol Secure)

> FTP Secure is used to securely transfer data between users, companies or clients. It's an extension of the File Transfer Protocol (FTP) which adds support for a security layer. FTPS uses cryptographic protocols like SSL (Secure Sockets Layer) or TLS (Transport Layer Security), although the former is now deprecated due to security vulnerabilities.

Reviewer: ```suggestion
FTP Secure is used to securely transfer data between computer systems. It's an extension of the File Transfer Protocol (FTP) which adds support for a security layer. FTPS uses cryptographic protocols like SSL (Secure Sockets Layer) or TLS (Transport Layer Security), although the former is now deprecated due to security vulnerabilities.

```
